### PR TITLE
get_gnss_information_sim70xx added ! (IDFGH-8118)

### PR DESCRIPTION
**Pull request by diplfranzhoepfinger**
3rd Try.

### CODE REVIEW

**Review by franz-ms-muc — APPROVED** [review]

**Comment by franz-ms-muc** [comment]
@david-cermak i think we are close. 
just need 1 Forward... (please help me) 
so i can comment out this line: 
https://github.com/espressif/esp-protocols/pull/119/commits/37cf5b1e009c6f70d91fbaa5e057234df791d520#diff-e5cf7c1d6038b9cbf36cb92e72665046de2974bfffaf5c4fa5bcd30fc1d9688dR184

**Comment by david-cermak** [comment]
> @david-cermak i think we are close. just need 1 Forward... (please help me) so i can comment out this line: [37cf5b1#diff-e5cf7c1d6038b9cbf36cb92e72665046de2974bfffaf5c4fa5bcd30fc1d9688dR184](https://github.com/espressif/esp-protocols/commit/37cf5b1e009c6f70d91fbaa5e057234df791d520#diff-e5cf7c1d6038b9cbf36cb92e72665046de2974bfffaf5c4fa5bcd30fc1d9688dR184)

The problem is that the `DCE` is make using `GenericDevice` module, and you'd need to create it with `SIM7070_gnss`.
It's possible to create a templated module `DCE_T<SIM7070_gnss>`, but without the comfort to call the AT commands directly.
It's also possible to create the default `DCE` and cast the generic device returned by `get_module()` to your `SIM7070_gnss`.

I've sketched up a commit that addresses both, but had to create a custom DCE from `DCE_T` of `SIM7070_gnss`, and repeat the AT command forwarding as it's done for the default `DCE_T<GenericDevice>`. PTAL:

[WIP-Make-custom-DCE-to-access-spec-module-api.patch.txt](https://github.com/espressif/esp-protocols/files/9529666/WIP-Make-custom-DCE-to-access-spec-module-api.patch.txt)

---
PS: I'm also fine with accepting the PR as is and address the issue separately by myself (as I understand it's not an easy arrangement...)

**Comment by diplfranzhoepfinger** [comment]
> > @david-cermak i think we are close. just need 1 Forward... (please help me) so i can comment out this line: [37cf5b1#diff-e5cf7c1d6038b9cbf36cb92e72665046de2974bfffaf5c4fa5bcd30fc1d9688dR184](https://github.com/espressif/esp-protocols/commit/37cf5b1e009c6f70d91fbaa5e057234df791d520#diff-e5cf7c1d6038b9cbf36cb92e72665046de2974bfffaf5c4fa5bcd30fc1d9688dR184)
> 
> The problem is that the `DCE` is make using `GenericDevice` module, and you'd need to create it with `SIM7070_gnss`. It's possible to create a templated module `DCE_T<SIM7070_gnss>`, but without the comfort to call the AT commands directly. It's also possible to create the default `DCE` and cast the generic device returned by `get_module()` to your `SIM7070_gnss`.
> 
> I've sketched up a commit that addresses both, but had to create a custom DCE from `DCE_T` of `SIM7070_gnss`, and repeat the AT command forwarding as it's done for the default `DCE_T<GenericDevice>`. PTAL:
> 
> [WIP-Make-custom-DCE-to-access-spec-module-api.patch.txt](https://github.com/espressif/esp-protocols/files/9529666/WIP-Make-custom-DCE-to-access-spec-module-api.patch.txt)
> 
> PS: I'm also fine with accepting the PR as is and address the issue separately by myself (as I understand it's not an easy arrangement...)

thanks a lot. With your help i was able to got it compiling. now must do a Test on the Board, but all looks fine for the Moment.

**Review by franz-ms-muc — APPROVED** [review]

**Review by franz-ms-muc — APPROVED** [review]

**Comment by diplfranzhoepfinger** [comment]
@david-cermak why this **Build and Run Target tests / build\_esp\_modem (release-v4.3, simple\_cmux\_client, esp32)** is failing ?

**Comment by diplfranzhoepfinger** [comment]
> @diplfranzhoepfinger Is this ready for review? Seems you've already fixed all the issues including the CI.

yes !

File: components/esp_modem/examples/simple_cmux_client/main/simple_cmux_client_main.cpp [hunk 1]
```diff
@@ -68,16 +83,33 @@ extern "C" void app_main(void)
     assert(esp_netif);
 
 #if CONFIG_EXAMPLE_MODEM_DEVICE_BG96 == 1
-    std::unique_ptr<DCE> dce = create_BG96_dce(&dce_config, dte, esp_netif);
+    auto dce = create_SIM7070_GNSS_dce(&dce_config, dte, esp_netif);
```
Comment (david-cermak) [inline components/esp_modem/examples/simple_cmux_client/main/simple_cmux_client_main.cpp:86]:
```suggestion
std::unique_ptr<DCE> dce = create_BG96_dce(&dce_config, dte, esp_netif);
```

Please revert this, or maybe
```patch
-std::unique_ptr<DCE> dce = create_BG96_dce(&dce_config, dte, esp_netif);
+auto dce = create_BG96_dce(&dce_config, dte, esp_netif);
```

File: components/esp_modem/examples/simple_cmux_client/main/simple_cmux_client_main.cpp [hunk 2]
```diff
@@ -68,16 +83,33 @@ extern "C" void app_main(void)
...
 #elif CONFIG_EXAMPLE_MODEM_DEVICE_SIM7600 == 1
-    std::unique_ptr<DCE> dce = create_SIM7600_dce(&dce_config, dte, esp_netif);
+    auto dce = create_SIM7600_dce(&dce_config, dte, esp_netif);
 #else
 #error "Unsupported device"
 #endif
     assert(dce);
 
+    if(dte_config.uart_config.flow_control == ESP_MODEM_FLOW_CONTROL_HW)
+    {
```
Comment (david-cermak) [inline components/esp_modem/examples/simple_cmux_client/main/simple_cmux_client_main.cpp:103]: Formatting
```suggestion
    if(dte_config.uart_config.flow_control == ESP_MODEM_FLOW_CONTROL_HW) {
```

File: components/esp_modem/include/cxx_include/command_library.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+namespace esp_modem::dce_commands {
+command_result generic_command(CommandableIf *t, const std::string &command,
+                               const std::list<std::string_view> &pass_phrase,
```
Comment (david-cermak) [inline components/esp_modem/include/cxx_include/command_library.hpp:18]: We don't document it anywhere but trying to comply with c++11 in public headers (but okay to use c++17 in sources).

This is helpful if people use the library with older SDKs (I think this won't compile with IDF `v4.3`)

**Review by david-cermak — APPROVED** [review]
left some nitpicks, but LGTM in general.

I think that we can merge it at this point, and I'd address the issues in a followup PR.

**Comment by david-cermak** [comment]
Accepted via https://github.com/espressif/esp-protocols/pull/165, merged as 652314e73d72f72b982d0b63d1f6748585c7bc91

**Pull request closed without merging**